_drafts/Firmware-encryption-with-python.md

```diff
@@ -0,0 +1,643 @@
+
+```python
+class  EncryptionCTR:
+    def __init__(self, key, iv ):
+        self.key = binascii.unhexlify(key.replace(":", ""))
+        self.iv  = binascii.unhexlify(iv.replace(":", ""))
+        self.count_ctr = Counter.new(112, prefix=self.iv, little_endian=False, initial_value=0) 
+        self.cipherObject = AES.new(self.key, AES.MODE_CTR, counter=self.count_ctr)
+    def EncryptCtr(self, plaintext):
+        return self.cipherObject.encrypt(bytes(plaintext))
+    def SwapLittleEndian(self,plaintext):
+        temp_swap = [bytes]*BYTES_READ
+        temp_swap[0] = plaintext[3]
+        temp_swap[1] = plaintext[2]
+        temp_swap[2] = plaintext[1]
+        temp_swap[3] = plaintext[0]
+        return temp_swap
+```
+The script needs the name of the app.bin firmware binary file to be encrypted and a name for a new file that corresponds to the encrypted firmware Cipherapp.bin. A 4-byte read of the original plaintext firmware file is performed, the format exchange is performed, we encrypt the block and finally the ciphertext block is written to a new binary file:
+
+```python
+def main():
+    print(__name__)
+    
+    PlainFirmware  = BinaryFile(sys.argv[1],'rb')
+    CipherFirmware = BinaryFile(sys.argv[2],'wb') 
+    CrytoFirmware  = EncryptionCTR(mykey,myiv)
+    bytes_remaining = 0
+    bytes_so_far_cryted = 0
+    BinaryFileSize = PlainFirmware.SizeFile()
+    bytes_remaining = BinaryFileSize - bytes_so_far_cryted
+    print("Bytes to encrypt :{}", bytes_remaining )
+    while(bytes_remaining):
+        Plainbytes = PlainFirmware.ReadFile(BYTES_READ)
+        PlainbytesLittle = CrytoFirmware.SwapLittleEndian(Plainbytes)
+        Cipherbytes = CrytoFirmware.EncryptCtr(PlainbytesLittle)
+        CipherbytesLittle = CrytoFirmware.SwapLittleEndian(Cipherbytes)
+        CipherFirmware.WriteFile(CipherbytesLittle)
+        print( PlainbytesLittle )
+        print( CipherbytesLittle )
+        
+        bytes_so_far_cryted += BYTES_READ
+        bytes_remaining = BinaryFileSize - bytes_so_far_cryted
+        if( bytes_remaining <= 0 ): bytes_remaining = 0
+        print( bytes_remaining )
+        
+    PlainFirmware.CloseFile()
+    CipherFirmware.CloseFile()
+main()
+```
+## Firmware Decryption Implementation
+
+Our update system managed to receive the encrypted firmware now it all depends on our MCU we have to decrypt the firmware and install it in a section of flash memory where it can be executed by our bootloader. In this post we are going to use the AES Hardware Accelerator to perform firmware decryption. All the code for this post is written for the [B-L4S5I-IOT01A](https://www.st.com/en/evaluation-tools/b-l4s5i-iot01a.html#) Dev kit with STM32L4S5 MCU by ST Micro.
```

Comment:
Might be good to touch on the fact that the AES Accelerator is a part on the STM32. The mention here is a bit vague.